zebra-chain/src/work/equihash.rs

```diff
@@ -60,11 +65,120 @@ impl Solution {
         Ok(())
     }
 
-    #[cfg(feature = "getblocktemplate-rpcs")]
+    /// Returns a [`Solution`] containing the bytes from `solution`.
+    /// Returns an error if `solution` is the wrong length.
+    pub fn from_bytes(solution: &[u8]) -> Result<Self, SerializationError> {
+        if solution.len() != SOLUTION_SIZE {
+            return Err(SerializationError::Parse(
+                "incorrect equihash solution size",
+            ));
+        }
+
+        let mut bytes = [0; SOLUTION_SIZE];
+        // Won't panic, because we just checked the length.
+        bytes.copy_from_slice(solution);
+
+        Ok(Self(bytes))
+    }
+
     /// Returns a [`Solution`] of `[0; SOLUTION_SIZE]` to be used in block proposals.
+    #[cfg(feature = "getblocktemplate-rpcs")]
     pub fn for_proposal() -> Self {
         Self([0; SOLUTION_SIZE])
     }
+
+    /// Mines and returns one or more [`Solution`]s based on a template `header`.
+    /// The returned header contains a valid `nonce` and `solution`.
+    ///
+    /// If `cancel_fn()` returns an error, returns early with `Err(SolverCancelled)`.
+    ///
+    /// The `nonce` in the header template is taken as the starting nonce. If you are running multiple
+    /// solvers at the same time, start them with different nonces.
+    /// The `solution` in the header template is ignored.
+    ///
+    /// This method is CPU and memory-intensive. It uses 144 MB of RAM and one CPU core while running.
+    /// It can run for minutes or hours if the network difficulty is high.
+    #[cfg(feature = "internal-miner")]
+    #[allow(clippy::unwrap_in_result)]
+    pub fn solve<F>(mut header: Header, mut cancel_fn: F) -> Result<Header, SolverCancelled>
+    where
+        F: FnMut() -> Result<(), SolverCancelled>,
+    {
+        use crate::shutdown::is_shutting_down;
+
+        let mut input = Vec::new();
+        header
+            .zcash_serialize(&mut input)
+            .expect("serialization into a vec can't fail");
+        let input = &input[0..Solution::INPUT_LENGTH];
+
+        while !is_shutting_down() {
+            // Don't run the solver if we'd just cancel it anyway.
+            cancel_fn()?;
+
+            let solutions = equihash_solver::tromp::solve_200_9_compressed(input, || {
+                // Cancel the solver if we have a new template.
+                if cancel_fn().is_err() {
+                    return None;
+                }
+
```

Comment:
Optional: 

```rust

/// Decrements `byte_array` by 1, interpreting it as a big-endian integer.
/// If the big-endian integer overflowed, sets all the bytes to one, and returns `true`.
pub fn decrement_big_endian(byte_array: &mut [u8]) -> bool {
    // Decrement the last byte in the array that is less than u8::MAX, and clear any bytes after it
    // to decrement the next value in big-endian (lexicographic) order.
    let is_wrapped_overflow = byte_array.iter_mut().rev().all(|v| {
        *v = v.wrapping_sub(1);
        v == &u8::MAX
    });

    is_wrapped_overflow
}

/// Modifies `nonce` to be the previous integer in big-endian order.
/// Wraps to max value if the previous nonce would overflow.
#[cfg(feature = "internal-miner")]
fn prev_nonce(nonce: &mut [u8; 32]) {
    let _ignore_overflow = crate::primitives::byte_array::decrement_big_endian(&mut nonce[..]);
}

// above the while loop:
let mut last_nonce = header.nonce;
Self::prev_nonce(&mut last_nonce);
```

```suggestion
                if header.nonce == last_nonce {
                    return None;
                }

```

Reply:
What does this change do, and why is it useful?
I can't evaluate your suggestion without that.

Also, suggestions that are this complex are better in another PR.

Reply:
> What does this change do

Hypothetically, the idea was to avoid retrying nonces until there's a new template.

Practically, nothing at all.

> and why is it useful?

It's not, I hadn't thought it through in context.

Reply:
Yeah given the size of the nonce space, and that they're arbitrary, it really doesn't matter if we skip a few.

That's why each solver gets initial nonces like:
- [0, 0, ... , 0, 1]
- [1, 0, ... , 0, 2]
- ...
- [254, 0, ... , 0, 255]
- [255, 0, ... , 0, 0]

It's a mess in the data but it's easy to code.

Reply:
Just curious, why change the last byte?

Reply:
As I wrote in the code comments, I don't want to have to care if the endianness of `next_nonce()` changes, particularly because it's defined in another crate.

It's also helpful to have a record of which solver thread mined each block for testing:
> We should check that the threads are actually doing different work, they should have nonces with large gaps between them.
> 
> the mining threads are running, and there are approximately the configured number of threads
> the [block nonces](https://testnet.zcashblockexplorer.com/blocks/000365cd36db8fcee91bbe4de75eac1a191ee5ede9a915f6f41fb636d7aec9f3) match [the pattern used by Zebra](https://github.com/ZcashFoundation/zebra/pull/8136/files#diff-d73b289b9e87396ee873d64e4bdd5f1d9e57f820adf73df42be03040a3341e89R487-R491), and it's not just the first thread successfully mining